Packs/PaloAltoNetworksCortex_v2/Integrations/PaloAltoNetworksCortex_v2/PaloAltoNetworksCortex_v2.py

```diff
@@ -0,0 +1,786 @@
+    Should only do requests and return data.
+    """
+
+    def __init__(self):
+        headers = {
+            'Authorization': REGISTRATION_ID,
+            'Accept': 'application/json'
+        }
+        super().__init__(base_url=TOKEN_RETRIEVAL_URL, headers=headers, verify=USE_SSL, proxy=PROXY)
+
+    def get_access_token(self):
+        integration_context = demisto.getIntegrationContext()
+        access_token = integration_context.get(ACCESS_TOKEN_CONST)
+        stored = integration_context.get('stored')
+        if access_token and stored:
+            if int(time.time()) - stored < 60 * 60 - 30:
+                return access_token
+        oproxy_response = self._http_request("POST",
+                                             "/cdl-token",
+                                             json_data={'token': get_encrypted(REFRESH_TOKEN, ENC_KEY)})
+        access_token = oproxy_response.get(ACCESS_TOKEN_CONST)
+        api_url = oproxy_response.get('url')
+        refresh_token = oproxy_response.get(REFRESH_TOKEN_CONST)
+        instance_id = oproxy_response.get(INSTANCE_ID_CONST)
+        updated_integration_context = {
+            ACCESS_TOKEN_CONST: access_token,
+            STORED_CONST: int(time.time()),
+            API_URL_CONST: api_url,
+            INSTANCE_ID_CONST: instance_id
+        }
+        if refresh_token:
+            updated_integration_context.update({REFRESH_TOKEN: refresh_token})
+        demisto.setIntegrationContext(updated_integration_context)
+        return access_token
+
+    def query_loggings(self, query):
+        """
+        This function handles all the querying of Cortex Logging service
+        """
+        query_data = {'query': self.add_instance_id_to_query(query),
+                      'language': 'bigquery'}
+        query_service = self.initial_query_service()
+        response = query_service.create_query(query_params=query_data)
+        query_result = response.json()
+
+        if not response.ok:
+            status_code = response.status_code
+            message = query_result[-1].get('context')
+            raise Exception(f'Error in query to Cortex [{status_code}] - {message}')
+
+        try:
+            results = [r.json() for r in query_service.iter_job_results(job_id=query_result.get("jobId"),
+                                                                        result_format='valuesDictionary',
+                                                                        max_wait=2000)]
+        except Exception as e:
+            raise Exception('Received error %s when querying logs.' % e)
+
+        extended_results: List[Dict] = []
+        for result in results:
+            page = result.get("page", {})
+            data = page.get("result", {}).get("data", [])
+            if data:
+                extended_results.extend(data)
+        records = [self.add_unique_record_id(record) for record in extended_results]
```

Comment:
what exactly are we doing here? assigning our own ID ?

Reply:
Yes, Talked to @jtschichold about this, there is no id given by the QueryService, and there must be an ID else the context will take only the last. He offered to con-cat the serial number of the firewall + time_generated + session_id and creating the id out of them, I thought it's a better idea to hash it because not all commands have these fields and it's easier to maintain.